xskillscore/core/probabilistic.py

```diff
@@ -550,48 +568,120 @@ def rps(
     * C. A. T. Ferro. Fair scores for ensemble forecasts. Q.R.J. Meteorol. Soc., 140:
       1917–1923, 2013. doi: 10.1002/qj.2270.
     * https://www-miklip.dkrz.de/about/problems/
+
     """
     bin_names = ["category"]
-    M = forecasts[member_dim].size
     bin_dim = f"{bin_names[0]}_bin"
-    # histogram(dim=[]) not allowed therefore add fake member dim
-    # to apply over when multi-dim observations
-    if len(observations.dims) == 1:
-        observations = histogram(
-            observations, bins=[category_edges], bin_names=bin_names, dim=None
-        )
-    else:
-        observations = histogram(
-            observations.expand_dims(member_dim),
+    M = forecasts[member_dim].size
+
+    def _bool_to_int(ds):
+        """convert xr.object of dtype bool to int to evade:
+        TypeError: numpy boolean subtract, the `-` operator, is not supported"""
+
+        def _helper_bool_to_int(da):
+            if da.dtype == "bool":
+                da = da.astype("int")
+            return da
+
+        if isinstance(ds, xr.Dataset):
+            ds = ds.map(_helper_bool_to_int)
+        else:
+            ds = _helper_bool_to_int(ds)
+        return ds
+
+    forecasts = _bool_to_int(forecasts)
+
+    def _check_identical_xr_types(a, b):
+        if type(a) != type(b):
+            raise ValueError(
+                f"a and b must be same type, found {type(a)} and {type(b)}"
+            )
+        for d in [a, b]:
+            if not isinstance(d, (xr.Dataset, xr.DataArray)):
+                raise ValueError("inputs must be xr.DataArray or xr.Dataset")
+
+    def _check_bin_dim(ds, bin_dim):
+        """Assert that bin_dim is in ds. Try to guess and rename edges dimension."""
+        for d in ["quantile", "threshold", "edge"]:
+            if d in ds.dims and bin_dim not in ds.dims:
+                ds = ds.rename({d: bin_dim})
+        if bin_dim not in ds.dims:
+            raise ValueError(f"require {bin_dim} dimension, found {ds.dims}")
+        return ds
+
+    _check_identical_xr_types(observations, forecasts)
+
+    # different ways of calculating RPS based on category_edges
+    if isinstance(category_edges, (xr.Dataset, xr.DataArray)) or isinstance(
+        category_edges, tuple
+    ):
+        if isinstance(
+            category_edges, tuple
+        ):  # edges tuple of two: use for obs and forecast edges separately
+            observations_edges, forecast_edges = category_edges
+            _check_identical_xr_types(forecast_edges, forecasts)
+            _check_identical_xr_types(observations_edges, forecasts)
+        else:  # edges only given once, so use for both obs and forecasts
+            _check_identical_xr_types(category_edges, forecasts)
+            observations_edges, forecast_edges = category_edges, category_edges
+
+        # cumulative probs
+        Fc = (forecasts < forecast_edges).mean(member_dim)
+        Oc = observations < observations_edges
+        # todo: mask Fc and Oc where all nans
```

Comment:
curious what this line is. want to create an issue for it? Can leave the comment in the code

Reply:
need to fix this. not a huge problem, but all nans should stay nans and not become 0.

Reply:
fixed now